pyomo/core/base/indexed_component.py

```diff
@@ -447,89 +490,190 @@ def itervalues(self):
         """Return an iterator of the component data objects in the dictionary"""
         for key in self:
             yield self[key]
-    
+
     def iteritems(self):
         """Return an iterator of (index,data) tuples from the dictionary"""
         for key in self:
             yield key, self[key]
 
-    def __getitem__(self, ndx):
+    def __getitem__(self, index):
         """
         This method returns the data corresponding to the given index.
         """
+        if self._constructed is False:
+            self._not_constructed_error(index)
+
         try:
-            _validIndex = ndx in self._data
+            obj = self._data.get(index, None)
         except TypeError:
-            # Process alternatives
-            return self._processUnhashableIndex(ndx, sys.exc_info()[1])
+            obj = None
+            index = self._processUnhashableIndex(index)
+            if index.__class__ is _IndexedComponent_slicer:
+                return index
+
+        if obj is None:
+            # Not good: we have to defer this import to now
+            # due to circular imports (expr imports _VarData
+            # imports indexed_component, but we need expr
+            # here
+            from pyomo.core.base import expr as EXPR
```

Comment:
With pyomo5, you won't have to defer this import.   You'll be able to add:

  from pyomo.core.expr import current as EXPR